File: pixie/storage/serialization.py
"""Span serialization and deserialization for database storage.

Handles conversion between frozen dataclass span instances and dict
representations suitable for Piccolo table rows, including nested types,
tuples, and datetimes.
"""

from __future__ import annotations

from dataclasses import fields
from datetime import datetime, timezone
from typing import Any

from pixie.instrumentation.spans import (
    AssistantMessage,
    ImageContent,
    LLMSpan,
    ObserveSpan,
    SystemMessage,
    TextContent,
    ToolCall,
    ToolDefinition,
    ToolResultMessage,
    UserMessage,
)


def serialize_span(span: ObserveSpan | LLMSpan) -> dict[str, Any]:
    """Convert a span to a dict matching Observation table columns.

    Returns a dict with keys: ``id``, ``trace_id``, ``parent_span_id``,
    ``span_kind``, ``name``, ``data``, ``error``, ``started_at``,
    ``ended_at``, ``duration_ms``.
    """
    data = _dataclass_to_dict(span)

    if isinstance(span, LLMSpan):
        return {
            "id": span.span_id,
            "trace_id": span.trace_id,
            "parent_span_id": span.parent_span_id,
            "span_kind": "llm",
            "name": span.request_model,
            "data": data,
            "error": span.error_type,
            "started_at": span.started_at,
            "ended_at": span.ended_at,
            "duration_ms": span.duration_ms,
        }

    # ObserveSpan
    return {
        "id": span.span_id,
        "trace_id": span.trace_id,
        "parent_span_id": span.parent_span_id,
        "span_kind": "observe",
        "name": span.name,
        "data": data,
        "error": span.error,
        "started_at": span.started_at,
        "ended_at": span.ended_at,
        "duration_ms": span.duration_ms,
    }


def _dataclass_to_dict(obj: Any) -> Any:
    """Recursively convert a dataclass to a JSON-safe dict.

    - Frozen dataclass → dict (via field introspection, not ``asdict``)
    - ``tuple`` → ``list``
    - ``datetime`` → ISO 8601 string
    - Other primitives pass through unchanged.
    """
    if obj is None:
        return None
    if isinstance(obj, datetime):
        return obj.isoformat()
    if isinstance(obj, (str, int, float, bool)):
        return obj
    if isinstance(obj, tuple):
        return [_dataclass_to_dict(item) for item in obj]
    if isinstance(obj, list):
        return [_dataclass_to_dict(item) for item in obj]
    if isinstance(obj, dict):
        return {k: _dataclass_to_dict(v) for k, v in obj.items()}
    # Dataclass
    if hasattr(obj, "__dataclass_fields__"):
        return {f.name: _dataclass_to_dict(getattr(obj, f.name)) for f in fields(obj)}
    return obj  # pragma: no cover


def deserialize_span(row: dict[str, Any]) -> ObserveSpan | LLMSpan:
    """Reconstruct a span from a table row dict.

    The ``span_kind`` field determines the target type. The ``data`` field
    contains the full nested payload.
    """
    kind = row["span_kind"]
    data: dict[str, Any] = row["data"]

    if kind == "llm":
        return _deserialize_llm(data)
    return _deserialize_observe(data)


def _deserialize_observe(data: dict[str, Any]) -> ObserveSpan:
    """Reconstruct an ObserveSpan from serialized data."""
    return ObserveSpan(
        span_id=data["span_id"],
        trace_id=data["trace_id"],
        parent_span_id=data["parent_span_id"],
        started_at=_parse_datetime(data["started_at"]),
        ended_at=_parse_datetime(data["ended_at"]),
        duration_ms=data["duration_ms"],
        name=data["name"],
        input=data["input"],
        output=data["output"],
        metadata=data["metadata"],
        error=data["error"],
    )


def _deserialize_llm(data: dict[str, Any]) -> LLMSpan:
    """Reconstruct an LLMSpan from serialized data."""
    return LLMSpan(
        span_id=data["span_id"],
        trace_id=data["trace_id"],
        parent_span_id=data["parent_span_id"],
        started_at=_parse_datetime(data["started_at"]),
        ended_at=_parse_datetime(data["ended_at"]),
        duration_ms=data["duration_ms"],
        operation=data["operation"],
        provider=data["provider"],
        request_model=data["request_model"],
        response_model=data["response_model"],
        input_tokens=data["input_tokens"],
        output_tokens=data["output_tokens"],
        cache_read_tokens=data["cache_read_tokens"],
        cache_creation_tokens=data["cache_creation_tokens"],
        request_temperature=data["request_temperature"],
        request_max_tokens=data["request_max_tokens"],
        request_top_p=data["request_top_p"],
        finish_reasons=tuple(data["finish_reasons"]),
        response_id=data["response_id"],
        output_type=data["output_type"],
        error_type=data["error_type"],
        input_messages=tuple(_deserialize_message(m) for m in data["input_messages"]),
        output_messages=tuple(_deserialize_assistant_message(m) for m in data["output_messages"]),
        tool_definitions=tuple(
            _deserialize_tool_definition(td) for td in data["tool_definitions"]
        ),
    )


# ── Message deserialization ───────────────────────────────────────────────────


def _deserialize_message(
    m: dict[str, Any],
) -> SystemMessage | UserMessage | AssistantMessage | ToolResultMessage:
    """Dispatch on ``role`` to reconstruct the right message type."""
    role = m["role"]
    if role == "system":
        return SystemMessage(content=m["content"])
    if role == "user":
        return UserMessage(
            content=tuple(_deserialize_content(c) for c in m["content"]),
        )
    if role == "assistant":
        return _deserialize_assistant_message(m)
    if role == "tool":
        return ToolResultMessage(
            content=m["content"],
            tool_call_id=m.get("tool_call_id"),
            tool_name=m.get("tool_name"),
        )
    raise ValueError(f"Unknown message role: {role}")  # pragma: no cover


def _deserialize_assistant_message(m: dict[str, Any]) -> AssistantMessage:
    """Reconstruct an AssistantMessage with content and tool calls."""
    return AssistantMessage(
        content=tuple(_deserialize_content(c) for c in m["content"]),
        tool_calls=tuple(_deserialize_tool_call(tc) for tc in m["tool_calls"]),
        finish_reason=m.get("finish_reason"),
    )


def _deserialize_content(c: dict[str, Any]) -> TextContent | ImageContent:
    """Dispatch on ``type`` to reconstruct the right content type."""
    ctype = c["type"]
    if ctype == "text":
        return TextContent(text=c["text"])
    if ctype == "image":
        return ImageContent(url=c["url"], detail=c.get("detail"))
    raise ValueError(f"Unknown content type: {ctype}")  # pragma: no cover


def _deserialize_tool_call(tc: dict[str, Any]) -> ToolCall:
    """Reconstruct a ToolCall."""
    return ToolCall(
        name=tc["name"],
        arguments=tc["arguments"],
        id=tc.get("id"),
    )


def _deserialize_tool_definition(td: dict[str, Any]) -> ToolDefinition:
    """Reconstruct a ToolDefinition."""
    return ToolDefinition(
        name=td["name"],
        description=td.get("description"),
        parameters=td.get("parameters"),
    )


# ── Utilities ─────────────────────────────────────────────────────────────────


def _parse_datetime(value: str | datetime) -> datetime:
    """Parse an ISO 8601 string to a timezone-aware datetime."""
    if isinstance(value, datetime):
        return value
    dt = datetime.fromisoformat(value)
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt

File: ts/src/storage/serialization.ts
/**
 * Span serialization and deserialization for database storage.
 *
 * Handles conversion between readonly span interfaces and plain objects
 * suitable for SQLite table rows, including nested types and Date↔ISO
 * string conversion.
 */

import type {
  AssistantMessage,
  ImageContent,
  LLMSpan,
  Message,
  MessageContent,
  ObserveSpan,
  SystemMessage,
  TextContent,
  ToolCall,
  ToolDefinition,
  ToolResultMessage,
  UserMessage,
} from "../instrumentation/spans";

// ── Serialize ────────────────────────────────────────────────────────────────

/**
 * Convert a span to a dict matching the Observation table columns.
 */
export function serializeSpan(
  span: ObserveSpan | LLMSpan
): Record<string, unknown> {
  const data = toPlainObject(span);

  if ("operation" in span) {
    // LLMSpan
    return {
      id: span.spanId,
      trace_id: span.traceId,
      parent_span_id: span.parentSpanId,
      span_kind: "llm",
      name: span.requestModel,
      data,
      error: span.errorType,
      started_at: span.startedAt.toISOString(),
      ended_at: span.endedAt.toISOString(),
      duration_ms: span.durationMs,
    };
  }

  // ObserveSpan
  return {
    id: span.spanId,
    trace_id: span.traceId,
    parent_span_id: span.parentSpanId,
    span_kind: "observe",
    name: span.name,
    data,
    error: span.error,
    started_at: span.startedAt.toISOString(),
    ended_at: span.endedAt.toISOString(),
    duration_ms: span.durationMs,
  };
}

/**
 * Recursively convert an object to a JSON-safe plain object.
 *
 * - `Date` → ISO 8601 string
 * - `readonly` arrays → plain arrays
 * - Other primitives pass through unchanged.
 */
function toPlainObject(obj: unknown): unknown {
  if (obj === null || obj === undefined) return null;
  if (obj instanceof Date) return obj.toISOString();
  if (typeof obj === "string" || typeof obj === "number" || typeof obj === "boolean") {
    return obj;
  }
  if (Array.isArray(obj)) {
    return obj.map(toPlainObject);
  }
  if (typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj as Record<string, unknown>)) {
      result[k] = toPlainObject(v);
    }
    return result;
  }
  return obj;
}

// ── Deserialize ──────────────────────────────────────────────────────────────

/**
 * Reconstruct a span from a table row dict.
 */
export function deserializeSpan(
  row: Record<string, unknown>
): ObserveSpan | LLMSpan {
  const kind = row["span_kind"] as string;
  const data = row["data"] as Record<string, unknown>;

  if (kind === "llm") {
    return deserializeLlm(data);
  }
  return deserializeObserve(data);
}

function deserializeObserve(data: Record<string, unknown>): ObserveSpan {
  return {
    spanId: data["spanId"] as string,
    traceId: data["traceId"] as string,
    parentSpanId: (data["parentSpanId"] as string | null) ?? null,
    startedAt: parseDate(data["startedAt"]),
    endedAt: parseDate(data["endedAt"]),
    durationMs: data["durationMs"] as number,
    name: (data["name"] as string | null) ?? null,
    input: data["input"],
    output: data["output"],
    metadata: (data["metadata"] as Readonly<Record<string, unknown>>) ?? {},
    error: (data["error"] as string | null) ?? null,
  };
}

function deserializeLlm(data: Record<string, unknown>): LLMSpan {
  const inputMessages = (data["inputMessages"] as Record<string, unknown>[]) ?? [];
  const outputMessages = (data["outputMessages"] as Record<string, unknown>[]) ?? [];
  const toolDefs = (data["toolDefinitions"] as Record<string, unknown>[]) ?? [];
  const finishReasons = (data["finishReasons"] as string[]) ?? [];

  return {
    spanId: data["spanId"] as string,
    traceId: data["traceId"] as string,
    parentSpanId: (data["parentSpanId"] as string | null) ?? null,
    startedAt: parseDate(data["startedAt"]),
    endedAt: parseDate(data["endedAt"]),
    durationMs: data["durationMs"] as number,
    operation: data["operation"] as string,
    provider: data["provider"] as string,
    requestModel: data["requestModel"] as string,
    responseModel: (data["responseModel"] as string | null) ?? null,
    inputTokens: data["inputTokens"] as number,
    outputTokens: data["outputTokens"] as number,
    cacheReadTokens: data["cacheReadTokens"] as number,
    cacheCreationTokens: data["cacheCreationTokens"] as number,
    requestTemperature: (data["requestTemperature"] as number | null) ?? null,
    requestMaxTokens: (data["requestMaxTokens"] as number | null) ?? null,
    requestTopP: (data["requestTopP"] as number | null) ?? null,
    finishReasons,
    responseId: (data["responseId"] as string | null) ?? null,
    outputType: (data["outputType"] as string | null) ?? null,
    errorType: (data["errorType"] as string | null) ?? null,
    inputMessages: inputMessages.map(deserializeMessage),
    outputMessages: outputMessages.map(deserializeAssistantMessage),
    toolDefinitions: toolDefs.map(deserializeToolDefinition),
  };
}

// ── Message deserialization ──────────────────────────────────────────────────

function deserializeMessage(m: Record<string, unknown>): Message {
  const role = m["role"] as string;
  if (role === "system") {
    return { role: "system", content: m["content"] as string } as SystemMessage;
  }
  if (role === "user") {
    const content = (m["content"] as Record<string, unknown>[]).map(deserializeContent);
    return { role: "user", content } as UserMessage;
  }
  if (role === "assistant") {
    return deserializeAssistantMessage(m);
  }
  if (role === "tool") {
    return {
      role: "tool",
      content: m["content"] as string,
      toolCallId: (m["toolCallId"] as string | null) ?? null,
      toolName: (m["toolName"] as string | null) ?? null,
    } as ToolResultMessage;
  }
  throw new Error(`Unknown message role: ${role}`);
}

function deserializeAssistantMessage(
  m: Record<string, unknown>
): AssistantMessage {
  const content = (m["content"] as Record<string, unknown>[]).map(deserializeContent);
  const toolCalls = ((m["toolCalls"] as Record<string, unknown>[]) ?? []).map(
    deserializeToolCall
  );
  return {
    role: "assistant",
    content,
    toolCalls,
    finishReason: (m["finishReason"] as string | null) ?? null,
  };
}

function deserializeContent(
  c: Record<string, unknown>
): TextContent | ImageContent {
  const ctype = c["type"] as string;
  if (ctype === "text") {
    return { type: "text", text: c["text"] as string };
  }
  if (ctype === "image") {
    return {
      type: "image",
      url: c["url"] as string,
      detail: (c["detail"] as string | null) ?? null,
    };
  }
  throw new Error(`Unknown content type: ${ctype}`);
}

function deserializeToolCall(tc: Record<string, unknown>): ToolCall {
  return {
    name: tc["name"] as string,
    arguments: (tc["arguments"] as Record<string, unknown>) ?? {},
    id: (tc["id"] as string | null) ?? null,
  };
}

function deserializeToolDefinition(td: Record<string, unknown>): ToolDefinition {
  return {
    name: td["name"] as string,
    description: (td["description"] as string | null) ?? null,
    parameters: (td["parameters"] as Record<string, unknown> | null) ?? null,
  };
}

// ── Utilities ────────────────────────────────────────────────────────────────

function parseDate(value: unknown): Date {
  if (value instanceof Date) return value;
  if (typeof value === "string") {
    const d = new Date(value);
    if (isNaN(d.getTime())) {
      throw new Error(
        `Invalid ISO 8601 date string: ${value}. ` +
          "Expected format: YYYY-MM-DDTHH:mm:ss.sssZ"
      );
    }
    return d;
  }
  throw new Error(`Cannot parse date from: ${String(value)}`);
}

export type { MessageContent };
